File: native/Sources/XRDApp/AnalysisStore.swift
import Foundation
import XRDCore

/// Codable snapshot of the engine `DGResult` — the fields the app needs to
/// persist and to show in Compare / CSV / history without making the whole engine
/// type Codable.
struct DGSnapshot: Codable {
    var dgPercent: Double
    var dgSigma: Double?
    var peakCount: Int
    var graphiticXc: Double
    var graphiticW: Double
    var graphiticMu: Double
    var graphiticD: Double
    var turboXc: Double?
    var turboW: Double?
    var turboD: Double?
    var areaGraphitic: Double
    var areaTurbostratic: Double
    var dPrime: Double
    var lc: Double
    var y0: Double
    var twoThetaOffset: Double
    var fitR2: Double
    var methodName: String
    var wavelength: Double

    init(_ r: DGResult) {
        dgPercent = r.dgPercent; dgSigma = r.dgSigma; peakCount = r.peakCount
        graphiticXc = r.graphitic.xc; graphiticW = r.graphitic.w
        graphiticMu = r.graphitic.mu; graphiticD = r.graphitic.dSpacing
        turboXc = r.turbostratic?.xc; turboW = r.turbostratic?.w; turboD = r.turbostratic?.dSpacing
        areaGraphitic = r.areaFractionGraphitic; areaTurbostratic = r.areaFractionTurbostratic
        dPrime = r.dPrimeWeighted; lc = r.crystalliteLc; y0 = r.y0
        twoThetaOffset = r.twoThetaOffset; fitR2 = r.fitR2
        methodName = r.methodName; wavelength = r.wavelength
    }
}

/// One past analysis, kept so prior deconvolutions aren't lost when settings change.
struct HistoryEntry: Codable {
    var timestamp: Date
    var settings: DeconvSettings
    var dgPercent: Double?
    var note: String?
}

/// The persisted per-file analysis record (the sidecar payload).
struct AnalysisRecord: Codable {
    var schemaVersion = 1
    var sourceFileName: String
    var settings: DeconvSettings
    var result: DGSnapshot?
    var shiftOffset: Double?          // applied 2θ displacement, if any
    var flaggedForRedo = false        // user marked this scan for re-running
    var history: [HistoryEntry] = []
    var updatedAt = Date()
    var appVersion: String?
}

extension AnalysisRecord {
    enum CodingKeys: String, CodingKey {
        case schemaVersion, sourceFileName, settings, result, shiftOffset
        case flaggedForRedo, history, updatedAt, appVersion
    }
    /// Tolerant decoder so sidecars written by older builds still load — missing
    /// keys (e.g. `flaggedForRedo`, added later) fall back to defaults instead of
    /// failing the whole decode (which would silently drop a saved analysis).
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init(sourceFileName: (try? c.decode(String.self, forKey: .sourceFileName)) ?? "",
                  settings: (try? c.decode(DeconvSettings.self, forKey: .settings)) ?? DeconvSettings())
        schemaVersion = (try? c.decode(Int.self, forKey: .schemaVersion)) ?? 1
        result = try? c.decode(DGSnapshot.self, forKey: .result)
        shiftOffset = try? c.decode(Double.self, forKey: .shiftOffset)
        flaggedForRedo = (try? c.decode(Bool.self, forKey: .flaggedForRedo)) ?? false
        history = (try? c.decode([HistoryEntry].self, forKey: .history)) ?? []
        updatedAt = (try? c.decode(Date.self, forKey: .updatedAt)) ?? Date()
        appVersion = try? c.decode(String.self, forKey: .appVersion)
    }
}

/// Reads/writes a sidecar JSON next to each scan: `MyScan.xy.xrda.json`. The raw
/// `.xy` is never modified — it is itself the original-data backup. Auto-loaded
/// when a file is opened; saved (coalesced) when the analysis changes. Write
/// failures (e.g. a read-only location) are reported, not fatal.
enum AnalysisStore {
    static let maxHistory = 20

    static func sidecarURL(for url: URL) -> URL {
        URL(fileURLWithPath: url.path + ".xrda.json")
    }

    private static var encoder: JSONEncoder {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys]
        e.dateEncodingStrategy = .iso8601
        return e
    }
    private static var decoder: JSONDecoder {
        let d = JSONDecoder(); d.dateDecodingStrategy = .iso8601; return d
    }

    static func load(for url: URL) -> AnalysisRecord? {
        guard let data = try? Data(contentsOf: sidecarURL(for: url)) else { return nil }
        return try? decoder.decode(AnalysisRecord.self, from: data)
    }

    @discardableResult
    static func save(_ record: AnalysisRecord, for url: URL) -> Bool {
        guard let data = try? encoder.encode(record) else { return false }
        do { try data.write(to: sidecarURL(for: url), options: .atomic); return true }
        catch { return false }
    }

    private static let appVersion =
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String

    /// Build/update the record for a file: set current settings + result snapshot,
    /// append a history entry when the settings differ from the last one (capped),
    /// and persist. Returns whether the write succeeded.
    @discardableResult
    static func update(for url: URL, settings: DeconvSettings, result: DGResult?,
                       flaggedForRedo: Bool? = nil) -> Bool {
        var rec = load(for: url) ?? AnalysisRecord(sourceFileName: url.lastPathComponent, settings: settings)
        // Append history only when the deconvolution settings actually changed.
        if rec.history.last?.settings != settings {
            rec.history.append(HistoryEntry(timestamp: Date(), settings: settings,
                                            dgPercent: result?.dgPercent, note: settings.aiNote))
            if rec.history.count > maxHistory { rec.history.removeFirst(rec.history.count - maxHistory) }
        }
        rec.settings = settings
        rec.result = result.map(DGSnapshot.init)
        rec.shiftOffset = result.flatMap { $0.twoThetaOffset != 0 ? $0.twoThetaOffset : nil }
        if let f = flaggedForRedo { rec.flaggedForRedo = f }   // nil = preserve existing
        rec.updatedAt = Date()
        rec.appVersion = appVersion
        return save(rec, for: url)
    }
}

File: windows/XRDAnalyzer/Services/AnalysisStore.cs
using System.Reflection;
using System.Text.Json;
using XRDAnalyzer.Engine.Models;
using XRDAnalyzer.Models;

namespace XRDAnalyzer.Services;

/// Reads/writes a sidecar JSON next to each scan: `MyScan.xy.xrda.json`. The
/// raw `.xy` is never modified. Mirrors `AnalysisStore`
/// (native/Sources/XRDApp/AnalysisStore.swift) — same sidecar shape, so files
/// are cross-platform between the macOS and Windows apps.
public static class AnalysisStore
{
    private const int MaxHistory = 20;

    private static readonly string? AppVersion = Assembly.GetExecutingAssembly().GetName().Version?.ToString();

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true,
        Converters = { new Iso8601DateTimeOffsetConverter() },
    };

    public static string SidecarPath(string path) => path + ".xrda.json";

    public static AnalysisRecord? Load(string path)
    {
        var sidecar = SidecarPath(path);
        if (!File.Exists(sidecar)) return null;
        try
        {
            return JsonSerializer.Deserialize<AnalysisRecord>(File.ReadAllText(sidecar), JsonOptions);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    public static bool Save(AnalysisRecord record, string path)
    {
        try
        {
            File.WriteAllText(SidecarPath(path), JsonSerializer.Serialize(record, JsonOptions));
            return true;
        }
        catch (IOException)
        {
            return false;
        }
    }

    /// Build/update the record for a file: set current settings + result
    /// snapshot, append a history entry when the settings differ from the
    /// last one (capped), and persist. `flaggedForRedo` of null preserves the
    /// existing flag.
    public static bool Update(string path, DeconvSettings settings, DGResultDto? result, bool? flaggedForRedo = null)
    {
        var rec = Load(path) ?? new AnalysisRecord { SourceFileName = Path.GetFileName(path), Settings = settings };
        if (rec.History.Count == 0 || !rec.History[^1].Settings.Equals(settings))
        {
            rec.History.Add(new HistoryEntry
            {
                Timestamp = DateTimeOffset.UtcNow,
                Settings = settings.Clone(),
                DgPercent = result?.DgPercent,
                Note = settings.AiNote,
            });
            if (rec.History.Count > MaxHistory) rec.History.RemoveRange(0, rec.History.Count - MaxHistory);
        }
        rec.Settings = settings;
        rec.Result = result is null ? null : DgSnapshot.From(result);
        rec.ShiftOffset = result is { TwoThetaOffset: not 0 } ? result.TwoThetaOffset : null;
        if (flaggedForRedo is { } f) rec.FlaggedForRedo = f;
        rec.UpdatedAt = DateTimeOffset.UtcNow;
        rec.AppVersion = AppVersion;
        return Save(rec, path);
    }

    /// Persist just the yield masses (leaves settings/result/history intact).
    public static bool SaveYield(string path, YieldInputs yield)
    {
        var rec = Load(path) ?? new AnalysisRecord { SourceFileName = Path.GetFileName(path), Settings = new DeconvSettings() };
        rec.YieldInputs = yield;
        rec.UpdatedAt = DateTimeOffset.UtcNow;
        rec.AppVersion = AppVersion;
        return Save(rec, path);
    }
}
